### PR TITLE
Fix timezone Date issue.

### PR DESCRIPTION
**Pull request by sashadoes**
## Summary

Addresses issue:

- #5816 

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/components/GoogleChart.js [hunk 1]
```diff
@@ -29,6 +29,7 @@ import { Chart } from 'react-google-charts';
 import set from 'lodash/set';
 import cloneDeep from 'lodash/cloneDeep';
 import merge from 'lodash/merge';
+import { stringToDate } from '../util/date-range/string-to-date';
```
Comment (asvinb) [inline assets/js/components/GoogleChart.js:32]: @sashadoes Please move the import within `Internal dependencies`

File: assets/js/components/notifications/BannerNotification/index.js [hunk 2]
```diff
@@ -23,6 +23,7 @@ import PropTypes from 'prop-types';
 import classnames from 'classnames';
 import map from 'lodash/map';
 import { useMount, useMountedState } from 'react-use';
+import { stringToDate } from '../../../util/date-range/string-to-date';
```
Comment (asvinb) [inline assets/js/components/notifications/BannerNotification/index.js:26]: @sashadoes Same, let's move that import within `Internal dependencies`

File: assets/js/modules/adsense/util/index.js [hunk 3]
```diff
@@ -20,6 +20,7 @@
  * External dependencies
  */
 import each from 'lodash/each';
+import { stringToDate } from '../../../util/date-range/string-to-date';
```
Comment (asvinb) [inline assets/js/modules/adsense/util/index.js:23]: @sashadoes Please add a comment for "Internal dependencies" similar to what we have in other files.

File: assets/js/modules/analytics/util/data-mock.js [hunk 4]
```diff
@@ -25,6 +25,7 @@ import castArray from 'lodash/castArray';
 import { zip, from, Observable } from 'rxjs';
 import { map, reduce, take } from 'rxjs/operators';
 import { MODULES_ANALYTICS } from '../datastore/constants';
```
Comment (asvinb) [inline assets/js/modules/analytics/util/data-mock.js:27]: @sashadoes Please move the `MODULES_ANALYTICS` and `stringToDate` imports within `Internal dependencies`

File: assets/js/modules/search-console/util/data-mock.js [hunk 5]
```diff
@@ -24,6 +24,7 @@ import faker from 'faker';
 import md5 from 'md5';
 import { range } from 'rxjs';
 import { map, reduce, take } from 'rxjs/operators';
+import { stringToDate } from '../../../util/date-range/string-to-date';
```
Comment (asvinb) [inline assets/js/modules/search-console/util/data-mock.js:27]: Same, should be within `Internal dependencies`

**Review by asvinb — CHANGES_REQUESTED** [review]
@sashadoes I left some comments in the PR. Can you check them out? Thanks!

File: assets/js/modules/adsense/util/index.js [hunk 6]
```diff
@@ -21,12 +21,16 @@
...
+/*
+ * Internal dependencies
+ */
 export * from './is-zero-report';
 export * from './parsing';
 export * from './site-stats-data';
 export * from './status';
 export * from './validation';
 export * from './url';
+import { stringToDate } from '../../../util/date-range/string-to-date';
```
Comment (asvinb) [inline assets/js/modules/adsense/util/index.js:33]: @sashadoes Please move the import directly under Internal dependencies. Exports should not be under internal dependencies.

**Review by asvinb — CHANGES_REQUESTED** [review]

**Comment by asvinb** [comment]
@sashadoes I made a small tweak: https://github.com/google/site-kit-wp/pull/5880/commits/678b5c4db1ea2becb2f386b76a8dfa62ed9c0a22 Otherwise LGTM

**Review by asvinb — APPROVED** [review]

**Review by techanvil — CHANGES_REQUESTED** [review]
Nice work here @sashadoes.

I have noticed one more place where we _could_ pass a string into a Date constructor:

https://github.com/google/site-kit-wp/blob/bbe3ba6ff1e035146cc781887befdda9cc7021a5/assets/js/util/date-range/get-date-string.js#L30-L39

At the moment, we are only using this function with `Date` instances.

However we should probably update the function to use `stringToDate` when the passed `date` arg is a string, to avoid unexpected bugs slipping through in future.

Alternatively we could update the parameter type to be `Date` rather than `string|number|Date`, I am open to suggestion. What do you think?

**Comment by sashadoes** [comment]
Hi @techanvil. Thank you for this shout :)
I tried both approaches and decided to go with the last one making `getDateString` work with Date obj only. 
My top argument is that the func Signatures work better with monolithic arguments, especially as the utility function should be 'stupid', as well as that it aimed to work with Date and return a string (even from the name), plus there are no real cases in the code where we pass a string.

Please let me know if you are happy.

**Comment by techanvil** [comment]
> Hi @techanvil. Thank you for this shout :) I tried both approaches and decided to go with the last one making `getDateString` work with Date obj only. My top argument is that the func Signatures work better with monolithic arguments, especially as the utility function should be 'stupid', as well as that it aimed to work with Date and return a string (even from the name), plus there are no real cases in the code where we pass a string.
> 
> Please let me know if you are happy.

Thanks @sashadoes, sounds good to me! Just a couple of small tweaks and this should be good to go.

**Review by techanvil — CHANGES_REQUESTED** [review]

**Comment by sashadoes** [comment]
Hi @techanvil . Looking at tests more precisely reveals that `getDateString ` receives the date number (nit a Date instance) here eg -> https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/adsense/util/data-mock.js#L262 (typeof new Date().setDate() is 'number' ).

At the same time option 1 you suggest previously has to modify `stringToDate` a lot as it is designed only to work with year-month-day strings pattern. Overall, putting extra logic inside `stringToDate` to serve it for `getDateString ` feels a bit over. 

However, I feel we can rethink improvements to make it more explicit and robust, but not sure if it should be done within this ticket's scope. 

Please let me know your thoughts. Thank you.

**Comment by techanvil** [comment]
Thanks @sashadoes!

> Hi @techanvil . Looking at tests more precisely reveals that `getDateString ` receives the date number (nit a Date instance) here eg -> https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/adsense/util/data-mock.js#L262 (typeof new Date().setDate() is 'number' ).

Good spot there :+1: 

> At the same time option 1 you suggest previously has to modify `stringToDate` a lot as it is designed only to work with year-month-day strings pattern. Overall, putting extra logic inside `stringToDate` to serve it for `getDateString ` feels a bit over.

I should have been a bit clearer - what I was trying to say is we could modify `getDateString`, not `stringToDate`. Specifically we could modify `getDateString` so it _calls_ `stringToDate` when passed a string:

```js
export const getDateString = ( date ) => {
	const dateInstance =
		typeof date === 'string' ? stringToDate( date ) : new Date( date );

	invariant(
		isValidDateInstance( dateInstance ),
		INVALID_DATE_INSTANCE_ERROR
	);

	// etc.
};
```

So I guess we could either update the test in `adsense/util/data-mock.js` to pass in a `Date` instance instead of a `number`, or make the change suggested above to `getDateString`.

On balance I'd go with the change to `getDateString` but happy either way.

**Comment by sashadoes** [comment]
Thank you @techanvil ,🙏🏻 I see the idea. But with that solution
```js
export const getDateString = ( date ) => {
	const dateInstance =
		typeof date === 'string' ? stringToDate( date ) : new Date( date );

	invariant(
		isValidDateInstance( dateInstance ),
		INVALID_DATE_INSTANCE_ERROR
	);

	// etc.
};
```
We will have vulnerability for the strings that has a different format than year-month-day as `stringToDate` can work only with.
So we have to extend `stringToDate`  or place a rule alike: 

```js
export const getDateString = ( date ) => {
	// check format yyyy-m-d
	const checkStringFormat = new RegExp(
		'^d{4}-(0?[1-9]|1[012])-(0?[1-9]|[12][0-9]|3[01])$'
	);
	const dateInstance = checkStringFormat.test( date )
		? stringToDate( date )
		: new Date( date );

	invariant(
		isValidDateInstance( dateInstance ),
		INVALID_DATE_INSTANCE_ERROR
	);

//etc...
```

What do you think?

**Comment by techanvil** [comment]
Hi @sashadoes, thanks for your patience here. I should have been cognisant of the fact that `stringToDate` only works for that `YY-MM-DD` format. Sorry to have wasted a bit of time here but I guess it's all valuable learning one way or the other!

That being the case, I think the right thing to do would in fact to go back to the simplification of `getDateString` to only accept `Date` instances.

We can then update the usage of it in `assets/js/modules/search-console/util/data-mock.js` and `assets/js/modules/adsense/util/data-mock.js`.

This seems preferable to me than adding extra complexity that isn't actually needed. And it ensures we don't leave a latent bug waiting to happen. Does that sound OK to you?

**Comment by sashadoes** [comment]
@techanvil No problem at all. Thank you for your time and assistance here.
Changing to our final version now 😇

**Review by tofumatt — APPROVED** [review]
Looks good to me 👍🏻

File: assets/js/modules/adsense/util/data-mock.js [hunk 7]
```diff
@@ -249,19 +250,20 @@ export function getAdSenseMockResponse( args ) {
...
-		map( ( item ) =>
-			getDateString(
-				new Date( startDate ).setDate( startDate.getDate() + item )
-			)
-		),
+		map( ( item ) => {
+			const updatedMilliseconds = new Date( startDate ).setDate(
+				startDate.getDate() + item
+			);
+			return getDateString( new Date( updatedMilliseconds ) );
```
Comment (aaemnnosttv) [inline assets/js/modules/adsense/util/data-mock.js:265]: This part is missing `stringToDate` use.

File: assets/js/modules/search-console/util/data-mock.js [hunk 8]
```diff
@@ -69,19 +70,20 @@ export function getSearchConsoleMockResponse( args ) {
...
-		map( ( item ) =>
-			getDateString(
-				new Date( startDate ).setDate( startDate.getDate() + item )
-			)
-		),
+		map( ( item ) => {
+			const updatedMilliseconds = new Date( startDate ).setDate(
+				startDate.getDate() + item
+			);
+			return getDateString( new Date( updatedMilliseconds ) );
```
Comment (aaemnnosttv) [inline assets/js/modules/search-console/util/data-mock.js:85]: Similar to before, this part is missing `stringToDate` use.

**Review by aaemnnosttv — COMMENTED** [review]
A few minor details that seem to have been missed here.

**Comment by aaemnnosttv** [comment]
Nevermind, I misread the code and it's only copying a date instance rather than parsing a string, so this is all good 👍

**Pull request merged**